Freedesktop module

dcbaker opened the pull request:
This is a work in process to add a module for creating freedesktop data files. There's a bunch of them, people need them, and providing a built-in method for generating and installing them to the correct place that is tested and validated seems pretty dang useful.

This is not completely as-is, the .desktop file is pretty complete I think, but the appstream generation is super limited still. The tests work on my system, but probably fail somewhere (windows?). There's also limited documentation. I've added automatic validation generation to the targets as well, which I think is pretty slick.

I'm mostly hoping to get some feedback before I spend too much more time on this.

dcbaker commented:
@ximion, I started thinking about this again I saw your metainfo creator. I thought you might be interested.

ximion commented:
Looks nice! I'll probably be able to take a much closer look at the code later (just quickly skimmed through it).

> There's also limited documentation.

With that, I can most certainly help, especially in case of AppStream and most likely in case of other XDG specs. What information are you missing?

dcbaker commented:
> With that, I can most certainly help, especially in case of AppStream and most likely in case of other XDG specs. What information are you missing?

Oh, I meant *I* need to write meson documentation for the new code :)

ximion commented:
> Oh, I meant I need to write meson documentation for the new code :)

Haha, yeah, that makes sense - apparently I am so used to people complaining about XDG docs that I now assume that implicitly, even if it isn't the case :P

xclaesse commented on docs/markdown/Freedesktop-module.md:
````diff
@@ -0,0 +1,47 @@
+  datafile desktop_file(exe, ...)
+```
+
+Creates a an datafile that will be installed if the executable used to
````
typo: "a an"

xclaesse commented on docs/markdown/Freedesktop-module.md:
```diff
@@ -0,0 +1,47 @@
+
+- `categories` a meson list of categories for this program. These values will be joined into a single semicolon delimited string in the .desktop file. Meson does not validate this value in any way. If not provided the field is unset.
+
+- `install_name` The name of the .desktop file without the file extension. If this is unset the name of the value of `name` or the name of the executable will be used, in that order.
```
"the name of the value of `name`" sounds weird. Do you mean "the value of the `name` field or the name of the executable..." ?

xclaesse commented on docs/markdown/snippets/freedesktop_module.md:
```diff
@@ -0,0 +1,15 @@
+    name : 'UnicornAwesomeApp',
+    generic_name : 'Email',
+    ...
+)
```
Would be nice to have the example in the module documentation too.

xclaesse commented on docs/sitemap.txt:
```diff
@@ -49,6 +49,7 @@ index.md
 			Windows-module.md
 			Cuda-module.md
 			Kconfig-module.md
+			Freedesktop-module.md
```
I think it needs to be added in `docs/theme/extra/templates/navbar_links.html` too.

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+
+    @permittedKwargs({'name', 'generic_name', 'icon', 'categories',
+                      'actions', 'install_name', 'mimetype', 'comment',
+                      'dbus'})
```
dbus kwarg is missing from doc.

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+            raise MesonException('Only one positinal argument allowed.')
+
+        if isinstance(args[0], Disabler):
+            return ModuleReturnValue(None, [])
```
If an argument is a disabler the interpreter should not even call your method, are you sure this can happen?

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+        value = listify(kwargs.get(kkey, []))
+        if value:
+            if any(not isinstance(i, str) for i in value):
+                raise MesonException('"{}" must be a list of strings if defined'.format(kkey))
```
stringlistify() does that check for you IIRC.

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+        if not isinstance(exe, Executable):
+            raise MesonException('First argument must be an Executable')
+
+        name = kwargs.get('name', '')  # type: str
```
You don't validate that name is a string.

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+            for a in actions:
+                if not isinstance(a, dict):
+                    raise MesonException('"actions" must be a list of dicts if defined.')
+            values['Actions'] = ';'.join(a['action'] for a in actions)
```
You should validate that 'action' is in the dict and is a string.

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+
+        for action in actions:
+            if 'action' not in action:
+                raise MesonException('missing required "action" entry in action "{}"'.format(action))
```
You validate after having using it.

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+            if 'action' not in action:
+                raise MesonException('missing required "action" entry in action "{}"'.format(action))
+            if 'name' not in action:
+                raise MesonException('missing required "name" entry in action "{}"'.format(action))
```
Need to validate it's a string too.

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+            avalues = {}
+            avalues['Name'] = action['name']
+
+            exec_ = listify(action['exec'])
```
You don't validate that 'exec' is in the dict, and you could use stringlistify() to validate they are strings.

xclaesse commented on mesonbuild/modules/freedesktop.py:
```diff
@@ -0,0 +1,206 @@
+
+            config['Desktop Action {}'.format(action['action'])] = avalues
+
+        install_name = kwargs.get('install_name', name)
```
You don't validate it's a string.

xclaesse commented on mesonbuild/interpreter.py:
```diff
@@ -703,9 +703,12 @@ def get_custom_install_mode(self):
+        ObjectHolder.__init__(self, data, subproject)
+        self.methods.update({
+            'sources': self.sources_method,
+        })
```
Hmm, Data object is not documented. The only place it's currently used is the return value of install_data() that is documented as being void... I'm not against adding this method, but we need to add a section under "returned objects" of reference manual.

dcbaker replied:
The more I think about it, the more I think this should be a custom_target. We want it to be updated whenever any of the inputs are updated anyway.

xclaesse commented:
@dcbaker It's all just small nitpickings, overall it looks good to me. I haven't looked at appstream part since it's not documented, and I don't know what they are.

One general question I've got: are there any escaping we should do? Does it work when the install path of the exe has spaces for example?

xclaesse commented:
Also, desktop files can have many more fields, even unrestricted `X-` fields. e.g. from random file on my distro, I see `X-Ubuntu-Gettext-Domain`, `X-Flatpak-RenamedFrom`, `StartupNotify`. Should we provide a freeform dictionary  kwarg for them? Like `extra_entries: {'StartupNotify': 'true', ...}`.

Oh, and I see you hardcode `values['Type'] = 'Application'`, are there other types of desktop files we could want to support? Mostly out of curiosity, they can always be added later.

xclaesse commented:
> One general question I've got: are there any escaping we should do? Does it work when the install path of the exe has spaces for example?

Self replying: Yes there are quoting/escaping considerations we should take care of: https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s07.html.

btw, linking to the spec, at least in the implementation code, could be nice.

dcbaker commented:
> Oh, and I see you hardcode values['Type'] = 'Application', are there other types of desktop files we could want to support? Mostly out of curiosity, they can always be added later.

There are two other types currently, `link` and `directory`. I don't think that either make sense for meson, but maybe someone else has a different opinion.

dcbaker commented:
> Also, desktop files can have many more fields, even unrestricted X- fields. e.g. from random file on my distro, I see X-Ubuntu-Gettext-Domain, X-Flatpak-RenamedFrom, StartupNotify. Should we provide a freeform dictionary kwarg for them? Like extra_entries: {'StartupNotify': 'true', ...}.

I think we should support all of the official keys as builtin functionality, but `X-` keys, sure, lets just put those in a a free form dictionary. We can even enforce that they're `X-` fields.

dcbaker commented:
Updated to return a CustomTarget, also added a keyword to automatically generate tests. I think the .desktop stuff is pretty much complete baring any bugs in my code (There is a bug I found in the desktop validator and sent an MR for). The appstream function still needs work I think.

xclaesse commented on docs/markdown/Freedesktop-module.md:
```diff
@@ -0,0 +1,112 @@
+  generic_name : 'IDE',
+  actions : [
+    {
+      name : 'Import',
```
keys must be quoted strings here.

xclaesse replied:
Unless we introduce intermediary object like `action = fd.action('Import', exec : [...])`

xclaesse commented:
@dcbaker could you explain why you opted for a custom target instead of generating the file at configure time? Note that some modules uses helper script run at build time and they put them into mesonbuild/scripts/ so they can be invoked as `meson --internal scriptname args` instead of mangling it in the module py file. Also you could use pickle to serialise your data instead of json, probably faster and safer regarding escaping issues you could have.

dcbaker commented:
Because I want to be able to pass the location of the .desktop file to the appstream() method (it needs to point at a .desktop file). Maybe it could be a run_target instead of a custom_target?

xclaesse commented:
@dcbaker I mean you could generate the .desktop file right away during configure, instead of delaying to build time, no? It's pretty much a configure_file() kind of operation as far as I can tell.

dcbaker commented:
I said run_target, but I mean run_command. I just need a value to pass from the .desktop generator to something else.

xclaesse commented:
Why not just writing the file directly in FreedesktopModule.desktop_file() and return `mesonlib.File.from_built_file(self.subdir, output)` just like configure_file() does?

eli-schwartz commented:
As a matter of curiosity, why would one wish to use this instead of using some sort of generator program to create a file and check it into git? For a fully static file I wouldn't want to waste build times (even very minor quantities of it) on having meson convert it from one format to another.

And at least I don't really understand the benefit of generating a .desktop file with a trivial file format using keyword arguments to a meson function that just become lines of the form `kwargname=kwargval1;kwargval2;kwargval3;kwargval4;`.

xclaesse commented:
Not a blocker, can be added later, but a post_install() method would make sense, similar to https://github.com/mesonbuild/meson/pull/8272. As suggested here: https://github.com/mesonbuild/meson/issues/8268#issuecomment-771782087.

xclaesse commented:
Right, that issue has been discussed above already, https://github.com/mesonbuild/meson/pull/6971#issuecomment-613713458. I think it make sense to have kwarg for the most common keys, but add a way to add free form str:str dictionary to extend it, and probably not limit that to `X-`.

ximion commented:
A dictionary won't cut it though - many of the tags in MetaInfo files have attributes (such as versions, comparison attributes, types, ...) and even more are actually deeply nested. A dict is flat - unless you want to nest dictionaries, but that complexity will probably get out of hand rather quickly, and will probably be harder to read than just writing the data outright.

eli-schwartz commented:
I will reiterate and expound on my belief that meson is supposed to configure files based on *build configuration options*, compile source code, install data, orchestrate tool processing, etc. -- it is not supposed to natively implement some file format and generate data files by transpiling meson dictionaries into XML or ini files. And as a side effect, completely waste time during `meson configure` by creating files that are a textbook case of "something that should be checked into git".

If the highly simple Freedesktop .desktop file format needs special tools to create them without syntax errors, meson or Freedesktop should provide a FAQ encouraging people to use some independent tool that helps you do that and produces files suitable for, again, checking into git.

A meson Freedesktop module should limit itself to:
- providing post-install integration for update-desktop-database / update-mime-database
- providing `install_foo()` methods that know *where* to install a source file better than the generic `install_data()` does
- optionally, `install_foo()` can be really clever and add a `meson test` rule to run `desktop-file-validate` or `appstreamcli validate` etc.

Please, please, please, do NOT get meson into the business of becoming a source code authoring system.
